File: html-database-v2/scripts/add_grid_coordinates.py
#!/usr/bin/env python3
"""
Add grid coordinates to Find a Grave records JSON file.
Parses section and plot information to generate gridX and gridY coordinates.
"""

import json
import re
from pathlib import Path

def parse_section_to_grid(section_str):
    """Convert section string to grid Y coordinate (row).
    Examples: 'A' -> 1, 'B' -> 2, '139B' -> 139*26 + 2, etc.
    """
    if not section_str:
        return None
    
    section_str = str(section_str).strip().upper()
    
    # Remove "Section" prefix
    section_str = re.sub(r'^SECTION\s*', '', section_str, flags=re.IGNORECASE)
    section_str = re.sub(r'^SEC\s*', '', section_str, flags=re.IGNORECASE)
    
    # Handle numeric sections like "139B" -> row 139, sub-row B
    match = re.match(r'^(\d+)([A-Z])?$', section_str)
    if match:
        num = int(match.group(1))
        letter = match.group(2)
        if letter:
            # Convert letter to number (A=1, B=2, etc.)
            letter_num = ord(letter) - ord('A') + 1
            return num * 100 + letter_num
        return num * 100
    
    # Handle pure letter sections (A, B, C, etc.)
    if len(section_str) == 1 and section_str.isalpha():
        return ord(section_str) - ord('A') + 1
    
    # Handle numeric only
    if section_str.isdigit():
        return int(section_str) * 100
    
    # Default: use hash for unknown formats
    return hash(section_str) % 1000

def parse_plot_to_grid(plot_str):
    """Convert plot string to grid X coordinate (column).
    Examples: '1' -> 1, '3' -> 3, 'A1' -> 1, etc.
    """
    if not plot_str:
        return None
    
    plot_str = str(plot_str).strip()
    
    # Extract numbers from plot string
    numbers = re.findall(r'\d+', plot_str)
    if numbers:
        return int(numbers[0])
    
    # If no numbers, use hash
    return hash(plot_str) % 100

def add_grid_coordinates(input_file, output_file=None):
    """Add gridX and gridY coordinates to JSON records."""
    if output_file is None:
        output_file = input_file
    
    print(f"Reading {input_file}...")
    with open(input_file, 'r', encoding='utf-8') as f:
        records = json.load(f)
    
    print(f"Processing {len(records)} records...")
    updated = 0
    
    for i, record in enumerate(records):
        # Parse plot info if available
        plot_info = record.get('plot', None)
        section = None
        plot_number = None
        
        if plot_info:
            # Try to extract section and plot from plot string
            # Format: "section 139B space 3" or "Section A, Plot 1"
            section_match = re.search(r'section\s*([A-Z0-9]+)', str(plot_info), re.IGNORECASE)
            if section_match:
                section = section_match.group(1)
            
            plot_match = re.search(r'(?:plot|space)\s*([A-Z0-9]+)', str(plot_info), re.IGNORECASE)
            if plot_match:
                plot_number = plot_match.group(1)
        
        # Calculate grid coordinates
        gridY = parse_section_to_grid(section) if section else None
        gridX = parse_plot_to_grid(plot_number) if plot_number else None
        
        # Add grid coordinates to record
        if gridX is not None or gridY is not None:
            record['gridX'] = gridX
            record['gridY'] = gridY
            updated += 1
        
        # Also store parsed section and plot for reference
        if section:
            record['section'] = section
        if plot_number:
            record['plotNumber'] = plot_number
    
    print(f"Added grid coordinates to {updated} records")
    
    print(f"Writing to {output_file}...")
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(records, f, indent=4, ensure_ascii=False)
    
    print("Done!")
    return updated

if __name__ == '__main__':
    script_dir = Path(__file__).parent
    input_file = script_dir / 'findagrave_records_complete.json'
    
    if not input_file.exists():
        print(f"Error: {input_file} not found!")
        exit(1)
    
    add_grid_coordinates(input_file)

File: html-database-v2/scripts/add_grid_coordinates.js
// Add grid coordinates to Find a Grave records JSON file
// Run with: node add_grid_coordinates.js

const fs = require('fs');
const path = require('path');

function parseSectionToGrid(sectionStr) {
    if (!sectionStr) return null;
    
    sectionStr = String(sectionStr).trim().toUpperCase();
    
    // Remove "Section" prefix
    sectionStr = sectionStr.replace(/^SECTION\s*/i, '');
    sectionStr = sectionStr.replace(/^SEC\s*/i, '');
    
    // Handle numeric sections like "139B" -> row 139, sub-row B
    const match = sectionStr.match(/^(\d+)([A-Z])?$/);
    if (match) {
        const num = parseInt(match[1]);
        const letter = match[2];
        if (letter) {
            // Convert letter to number (A=1, B=2, etc.)
            const letterNum = letter.charCodeAt(0) - 'A'.charCodeAt(0) + 1;
            return num * 100 + letterNum;
        }
        return num * 100;
    }
    
    // Handle pure letter sections (A, B, C, etc.)
    if (sectionStr.length === 1 && /[A-Z]/.test(sectionStr)) {
        return sectionStr.charCodeAt(0) - 'A'.charCodeAt(0) + 1;
    }
    
    // Handle numeric only
    if (/^\d+$/.test(sectionStr)) {
        return parseInt(sectionStr) * 100;
    }
    
    // Default: use hash for unknown formats
    let hash = 0;
    for (let i = 0; i < sectionStr.length; i++) {
        hash = ((hash << 5) - hash) + sectionStr.charCodeAt(i);
        hash = hash & hash;
    }
    return Math.abs(hash) % 1000;
}

function parsePlotToGrid(plotStr) {
    if (!plotStr) return null;
    
    plotStr = String(plotStr).trim();
    
    // Extract numbers from plot string
    const numbers = plotStr.match(/\d+/);
    if (numbers) {
        return parseInt(numbers[0]);
    }
    
    // If no numbers, use hash
    let hash = 0;
    for (let i = 0; i < plotStr.length; i++) {
        hash = ((hash << 5) - hash) + plotStr.charCodeAt(i);
        hash = hash & hash;
    }
    return Math.abs(hash) % 100;
}

function addGridCoordinates(inputFile, outputFile) {
    if (!outputFile) outputFile = inputFile;
    
    console.log(`Reading ${inputFile}...`);
    const data = fs.readFileSync(inputFile, 'utf8');
    const records = JSON.parse(data);
    
    console.log(`Processing ${records.length} records...`);
    let updated = 0;
    
    for (let i = 0; i < records.length; i++) {
        const record = records[i];
        
        // Parse plot info if available
        const plotInfo = record.plot || null;
        let section = null;
        let plotNumber = null;
        
        if (plotInfo) {
            // Try to extract section and plot from plot string
            const sectionMatch = String(plotInfo).match(/section\s*([A-Z0-9]+)/i);
            if (sectionMatch) {
                section = sectionMatch[1];
            }
            
            const plotMatch = String(plotInfo).match(/(?:plot|space)\s*([A-Z0-9]+)/i);
            if (plotMatch) {
                plotNumber = plotMatch[1];
            }
        }
        
        // Calculate grid coordinates
        const gridY = section ? parseSectionToGrid(section) : null;
        const gridX = plotNumber ? parsePlotToGrid(plotNumber) : null;
        
        // Add grid coordinates to record
        if (gridX !== null || gridY !== null) {
            record.gridX = gridX;
            record.gridY = gridY;
            updated++;
        }
        
        // Also store parsed section and plot for reference
        if (section) {
            record.section = section;
        }
        if (plotNumber) {
            record.plotNumber = plotNumber;
        }
    }
    
    console.log(`Added grid coordinates to ${updated} records`);
    
    console.log(`Writing to ${outputFile}...`);
    fs.writeFileSync(outputFile, JSON.stringify(records, null, 4), 'utf8');
    
    console.log('Done!');
    return updated;
}

// Run if called directly
if (require.main === module) {
    const scriptDir = __dirname;
    const inputFile = path.join(scriptDir, 'findagrave_records_complete.json');
    
    if (!fs.existsSync(inputFile)) {
        console.error(`Error: ${inputFile} not found!`);
        process.exit(1);
    }
    
    addGridCoordinates(inputFile);
}

module.exports = { addGridCoordinates, parseSectionToGrid, parsePlotToGrid };

